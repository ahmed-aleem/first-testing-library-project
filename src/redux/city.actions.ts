import {cityActionTypes} from './city.types';
import Axios from 'axios';

const api_token = '67rZf932yce7QTVexA36lr0L3_gA91EaSh2aEUnVHVlUBc656B4uTSvn5zYUjvYP3h0';

export const fetchCitiesStart: Function = () => ({
    type: cityActionTypes.SET_CITIES_START
})

export const fetchCitiesSuccess = (cities: {city_name: string}[]) => ({
    type: cityActionTypes.SET_CIITIES_SUCCESS,
    payload: cities
})

export const fetchCitiesFailure = (message: string) => ({
    type: cityActionTypes.SET_CITIES_FAILURE,
    payload: message
})

export const fetchCitiesAsync = () => async (dispatch: Function) => {
    
    dispatch(fetchCitiesStart());

    const {data} = await Axios.get('https://www.universal-tutorial.com/api/getaccesstoken', {headers: {'Accept': 'application/json', 'api-token': api_token, 'user-email': 'ahmedatjs@gmail.com'}});

    const resultCities = await Axios.get(`https://www.universal-tutorial.com/api/cities/Telangana`, { headers: {'Authorization': `Bearer ${data.auth_token}`, 'Accept': 'application/json'} }); 

    console.log(resultCities, 'result');

    resultCities.data ? dispatch(fetchCitiesSuccess(resultCities.data)) : dispatch(fetchCitiesFailure(resultCities.statusText))
}

export const setSearch = (keyword: string) => ({
    type: cityActionTypes.SET_SEARCH_KEY,
    payload: keyword
})