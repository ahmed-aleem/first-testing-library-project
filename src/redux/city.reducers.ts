import {cityActionTypes} from './city.types';
import {getFilteredCities} from '../utils/filterData';

export interface BasicState{
    isCities: boolean,
    cities: [{city_name?: string}],
    filteredCities: [{city_name?: string}],
    searchKey: string | null,
    errorMessage: string | null
}

export interface BasicAction{
    type: string,
    payload: {city_name?: string}[] | string | Function
}

export const INITIAL_STATE: BasicState = {
    isCities: false,
    cities: [{}],
    filteredCities: [{}],
    searchKey: '',
    errorMessage: ''
}

const cityReducer = (state = INITIAL_STATE, action: BasicAction) => {

    switch (action.type) {
        case cityActionTypes.SET_CITIES_START:
            return {
                ...state,
                isCities: false
            };
            
        case cityActionTypes.SET_CIITIES_SUCCESS:
            return {
                ...state,
                isCities: true,
                cities: action.payload
            };

        case cityActionTypes.SET_CITIES_FAILURE:
            return {
                ...state,
                isCities: true,
                errorMessage: action.payload
            };
        case cityActionTypes.SET_SEARCH_KEY:
            return {
                ...state,
                searchKey: action.payload,
                filteredCities: getFilteredCities(state.cities, action.payload as string)
            };

        default:
            return state;
    }

}

export default cityReducer;