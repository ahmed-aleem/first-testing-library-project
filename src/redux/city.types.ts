export const cityActionTypes = {
    SET_CITIES_START: 'SET_CITIES_START',
    SET_CIITIES_SUCCESS: 'SET_CITIES_SUCCESS',
    SET_CITIES_FAILURE: 'SET_CITIES_FAILURE',
    SET_SEARCH_KEY: 'SET_SEARCH_KEY'
}