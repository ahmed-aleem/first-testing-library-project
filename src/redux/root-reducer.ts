import { combineReducers } from 'redux';
import cityReducer from './city.reducers';

const rootReducer = combineReducers({
    city: cityReducer
})

export type RootState = ReturnType<typeof rootReducer>

export default rootReducer;