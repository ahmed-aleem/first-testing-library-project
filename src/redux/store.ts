import {createStore, applyMiddleware, compose} from 'redux';
import thunk from 'redux-thunk';
import {composeWithDevTools} from 'redux-devtools-extension';

import rootReducer from './root-reducer';

const middlewares = [thunk];

const composeEnhancers = composeWithDevTools || compose;

const store = createStore(rootReducer, composeEnhancers(applyMiddleware(...middlewares)));

export default store;