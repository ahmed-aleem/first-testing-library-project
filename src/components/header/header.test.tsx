import React, { Props, ReactElement } from 'react';
import {createStore} from 'redux';
import {Provider} from 'react-redux';
import {render, cleanup} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Header from './header.component';
import rootReducer from '../../redux/root-reducer';

afterEach(cleanup);

function renderWithRedux( component: ReactElement, {store = createStore(rootReducer)} ={}){
    return {
    ...render(<Provider store={store}>{component}</Provider>)
    }
}

it('renders', () => {
    const {asFragment} = renderWithRedux(<Header />);
    expect(asFragment()).toMatchSnapshot();
})