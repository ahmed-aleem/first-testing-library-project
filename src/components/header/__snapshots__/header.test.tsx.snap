// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div
    class="header"
  >
    <h1>
      Welcome to Telangana!
    </h1>
    <p>
      Explore activities in 50+ cities across the state with us
    </p>
    <div
      class="search-section"
    >
      <input
        class="search-input"
        name="search"
        placeholder="search for a city in telangana"
        type="search"
        value=""
      />
      <div
        class="search-results-container"
      />
    </div>
  </div>
</DocumentFragment>
`;
